docs/manual/src/udl/decorators.md

```diff
@@ -0,0 +1,242 @@
+
+class MyDemoDecorator(
+    val backgroundScope: CoroutineContext,
+    val errorReporter: (e: Exception) -> Unit,
+    val listener: () -> Void
+) {
+    fun <T> onBackgroundThread(rustCall: () -> T) {
+        backgroundScope.launch {
+            rustCall()
+        }
+    }
+
+    fun <T> withErrorReporter(rustCall: () -> T) =
+        try {
+            rustCall()
+        } catch (e: Exception) {
+            errorReporter(e)
+        }
+
+    fun thenNotifyListeners(rustCall: () -> T) {
+        try {
+            rustCall()
+        } finally {
+            listener()
+        }
+    }
+}
+```
+
+### Specify the decorator class in your uniffi.toml
+
+```toml
+
+[bindings.kotlin.decorators.DemoRustObject]
+class_name = MyDemoDecorator
+```
+
+### Use the decorator
+
+The above code will make UniFFI define a `DemoLib` class that uses
+`MyDemoDecorator` to decorate calls to `DemoRustObject`.  The `DemoLib`
+constructor will input a `DemoRustObject` and `MyDemoDecorator`.
+
+For example:
+
+```kotlin
+val demoLib = DemoLib(
+   DemoRustObject(),
+   MyDemoDecorator(backgroundScope, errorReporter, listener))
+
+// This will call MyDemoDecorator.stateChange { DemoRustObject.doConsequentialThing() }
+demoLib.doConsequentialThing()
+```
+
+This is a considerable improvement! Now the decorator methods can be specified by the app, and the UDL uses them. Each time the UDL changes, the foreign language bindings keeps up.
+
+### Re-writing the wrapper
+
+This would happen differently than the original proposal, but I'm not sure
+exactly what to suggest because I don't understand the issue so well.  If the
+point is to avoid changes to the code that constructs DemoLib/DemoRustObject,
+then what about one of these?
+
+  - Making DemoRustObject an open class and DemoLib a subclass whose constructor inputs the args you want
+  - A factory function that inputs the args you want and returns a DemoRustObject
+
+## Specification
+
+### UDL
+
+* Specify that an interface can be decorated by using the `[Decorated={class_name}]` attribute
+  * `{class_name}` specifies that UniFFI should define a class named `class_name` where methods of the interface are called through decorators
+* Specify how each method will be decorated in the decorator class using the `[CallsWith={decorator_name}]` attribute
+  * This specifies that this method should be called with function named `decorator_name`
+  * Methods that don't have the `CallsWith` attribute will be called directly
+
+### uniffi.toml
+
+* To enable a decorated class for a specific bindings language, add a table named `[bindings.{language}.decorators.{interface_name}]` with the following keys:
+  * `class_name`: name of the decorator implementation class
+  * `import` (optional): import line to add at the top of the file.  Use this to import your decorator class if needed.
+  * `return_types` (swift only): Table that maps decorator names to the return type for the decorator function.
```

Comment:
First off, I agree that it would be nice to generate the interface on Kotlin, which means that the return type would need to be specified there as well.  I also think that would help with error messages, since you should get a message saying that you didn't implement the interface correctly.

On throwing: I think you could have a value in `uniffi.toml` that specifies the throwing behavior.  I would start with these options:
  - `rethrows` (the default).  The decorator may re-throw exceptions from the original function
  - `handles`.  The decorator handles exception from the original function and never throw.

On `await`/`async`:   I was thinking that rather than making the decorator `async` itself, it could return something that was compatible with `async` code.  For example, in Kotlin, the decorator can return a `Future<T>`, and a `suspend` function could call it's `.await()` method.  I'm not sure if that works in swift or not.

That said, I think this could be a config value as well.  If you specify `async=true` in `uniffi.toml`, then we could add `async` to the decorator method signature and the call site.